src/k8s/cmd/k8sd/k8sd_cluster_recover.go

```diff
@@ -0,0 +1,303 @@
+package k8sd
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path"
+	"regexp"
+	"slices"
+	"strings"
+
+	"github.com/canonical/go-dqlite"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/termios"
+	"github.com/canonical/microcluster/v2/cluster"
+	"github.com/canonical/microcluster/v2/microcluster"
+	"github.com/spf13/cobra"
+	"golang.org/x/sys/unix"
+	"gopkg.in/yaml.v2"
+)
+
+const recoveryConfirmation = `You should only run this command if:
+ - A quorum of cluster members is permanently lost
+ - You are *absolutely* sure all k8s daemons are stopped (sudo snap stop k8s)
+ - This instance has the most up to date database
+
+Do you want to proceed? (yes/no): `
+
+const clusterYamlRecoveryComment = `# Member roles can be modified. Unrecoverable nodes should be given the role "spare".
+#
+# "voter" - Voting member of the database. A majority of voters is a quorum.
+# "stand-by" - Non-voting member of the database; can be promoted to voter.
+# "spare" - Not a member of the database.
+#
+# The edit is aborted if:
+# - the number of members changes
+# - the name of any member changes
+# - the ID of any member changes
+# - the address of any member changes
+# - no changes are made
+`
+
+const infoYamlRecoveryComment = `# Verify the ID, address and role of the local node.
+#
+# Cluster members:
+`
+
+const daemonYamlRecoveryComment = `# Verify the name and address of the local node.
+#
+# Cluster members:
+`
+
+var clusterRecoverOpts struct {
+	K8sDqliteStateDir string
+}
+
+
+func newClusterRecoverCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "cluster-recover",
+		Short: "Recover the cluster from this member if quorum is lost",
+		RunE:  clusterRecover,
+	}
+
+	cmd.Flags().StringVar(&clusterRecoverOpts.K8sDqliteStateDir, "state-dir",
+						  "", "k8s-dqlite datastore location")
+
+	return cmd
+}
+
+func removeYamlComments(content []byte) []byte {
+	re := regexp.MustCompile("(?s)#.*?\n")
+	return re.ReplaceAll(content, nil)
+}
+
+func removeYamlCommentsFromFile(path string) error {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	content = removeYamlComments(content)
+
+	return os.WriteFile(path, content, os.FileMode(0o644))
```

Comment:
``os.WriteFile`` ignores the file mode if the file already exists, which it does since we've just opened it.

```
WriteFile creates it with permissions perm (before umask); otherwise WriteFile truncates it before writing, without changing permission
```